File: sandbox/backend/java/simple-eccomerce/src/main/java/com/simple/ecommerce/application/service/ShoppingCartService.java
package com.simple.ecommerce.application.service;

import com.simple.ecommerce.application.exception.EmptyCartException;
import com.simple.ecommerce.application.exception.ProductUnavailableException;
import com.simple.ecommerce.application.port.PromotionStrategy;
import com.simple.ecommerce.domain.entity.Product;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

/**
 * Service for managing shopping cart operations.
 * Business Rules:
 * - Initially empty
 * - Can have multiple units of same product
 * - Total price = sum of product prices with applied promotion
 * - Only one active promotion at a time
 * - New promotion overrides previous one
 */
public class ShoppingCartService {
    
    private final Map<Product, Integer> products;
    private PromotionStrategy activePromotion;

    /**
     * Creates an empty shopping cart service.
     */
    public ShoppingCartService() {
        this.products = new HashMap<>();
    }

    /**
     * Adds a product to the cart. If the product is already in the cart,
     * the quantity is increased by 1.
     *
     * @param product the product to add
     * @throws IllegalArgumentException if product is null
     * @throws ProductUnavailableException if product is unavailable
     */
    public void addProduct(Product product) {
        if (product == null) {
            throw new IllegalArgumentException("Product cannot be null");
        }
        
        if (!product.isAvailable()) {
            throw new ProductUnavailableException(product);
        }
        
        products.put(product, products.getOrDefault(product, 0) + 1);
    }

    /**
     * Removes a product from the cart. If the product has multiple units,
     * only one unit is removed.
     *
     * @param product the product to remove
     * @throws IllegalArgumentException if product is null
     */
    public void removeProduct(Product product) {
        if (product == null) {
            throw new IllegalArgumentException("Product cannot be null");
        }
        
        if (!products.containsKey(product)) {
            return; // Product not in cart, nothing to remove
        }
        
        int quantity = products.get(product);
        if (quantity > 1) {
            products.put(product, quantity - 1);
        } else {
            products.remove(product);
        }
    }

    /**
     * Returns the contents of the cart.
     *
     * @return an unmodifiable map of products and their quantities
     */
    public Map<Product, Integer> getCartContents() {
        return Collections.unmodifiableMap(products);
    }

    /**
     * Calculates the total price of all products in the cart,
     * applying any active promotion.
     *
     * @return the total price
     * @throws EmptyCartException if the cart is empty
     */
    public double calculateCartPrice() {
        if (products.isEmpty()) {
            throw new EmptyCartException("Cannot calculate price for an empty cart");
        }
        
        double totalPrice = products.entrySet().stream()
                .mapToDouble(entry -> entry.getKey().getPrice() * entry.getValue())
                .sum();
        
        if (activePromotion != null) {
            double discount = activePromotion.calculateDiscount(products);
            return Math.max(0, totalPrice - discount);
        }
        
        return totalPrice;
    }

    /**
     * Activates a promotion for this cart. Any previously active promotion is overridden.
     *
     * @param promotion the promotion to activate
     */
    public void activatePromotion(PromotionStrategy promotion) {
        this.activePromotion = promotion;
    }
    
    /**
     * Checks if the cart is empty.
     *
     * @return true if cart is empty, false otherwise
     */
    public boolean isEmpty() {
        return products.isEmpty();
    }
    
    /**
     * Clears all products from the cart.
     */
    public void clear() {
        products.clear();
    }
} 

File: Projects/Testtrack/p911 - Simple eCommerce backend/application/service/shopping_cart_service.py
from typing import Dict, Optional
from application.exception.empty_cart_exception import EmptyCartException
from application.exception.product_unavailable_exception import ProductUnavailableException
from application.port.promotion_strategy import PromotionStrategy
from domain import Product


class ShoppingCartService:
    def __init__(self):
        self.products: Dict[Product, int] = {}
        self.active_promotion: Optional[PromotionStrategy] = None

    def add_product(self, product: Product) -> None:
        if product is None:
            raise ValueError("Product cannot be null")
        if not product.is_available():
            raise ProductUnavailableException(product)

        self.products[product] = self.products.get(product, 0) + 1

    def remove_product(self, product: Product) -> None:
        if product is None:
            raise ValueError("Product cannot be null")
        if product not in self.products:
            return

        quantity = self.products[product]
        if quantity > 1:
            self.products[product] = quantity - 1
        else:
            del self.products[product]

    def get_cart_contents(self) -> Dict[Product, int]:
        return self.products.copy()

    def calculate_cart_price(self) -> float:
        if not self.products:
            raise EmptyCartException(
                "Cannot calculate price for an empty cart")

        total_price = sum(product.get_price() * quantity for product,
                          quantity in self.products.items())

        if self.active_promotion is not None:
            discount = self.active_promotion.calculate_discount(self.products)
            return max(0, total_price - discount)

        return total_price

    def activate_promotion(self, promotion: PromotionStrategy) -> None:
        self.active_promotion = promotion

    def is_empty(self) -> bool:
        return len(self.products) == 0

    def clear(self) -> None:
        self.products.clear()
